sdk/src/router/public/router-utils.ts

```diff
@@ -135,6 +139,63 @@ export class RouterUtils {
     return null;
   }
 
+  /**
+   * Calculate the price impact for a route.
+   * @param trade The trade the user used to derive the route.
+   * @param route The route to calculate the price impact for.
+   * @returns A Decimal object representing the percentage value of the price impact (ex. 3.01%)
+   */
+  static getPriceImpactForRoute(trade: Trade, route: TradeRoute): Decimal {
+    const { amountSpecifiedIsInput } = trade;
+    const directionalSubroutes = amountSpecifiedIsInput
+      ? route.subRoutes
+      : route.subRoutes.slice().reverse();
+
+    const totalBaseValue = directionalSubroutes.reduce((acc, route) => {
+      const directionalHops = amountSpecifiedIsInput
+        ? route.hopQuotes
+        : route.hopQuotes.slice().reverse();
+      const baseOutputs = directionalHops.reduce((acc, quote, index) => {
+        const { snapshot } = quote;
+        const { aToB, sqrtPrice, totalFeeRate } = snapshot;
+        // Inverse sqrt price will cause 1bps precision loss since ticks are spaces of 1bps
+        const directionalSqrtPrice = aToB ? sqrtPrice : PriceMath.invertSqrtPriceX64(sqrtPrice);
+
+        // Convert from in/out -> base_out/in using the directional price & fee rate
+        let nextBaseValue;
+        const price = directionalSqrtPrice.mul(directionalSqrtPrice).div(U64);
+        if (amountSpecifiedIsInput) {
+          const amountIn = index === 0 ? quote.amountIn : acc[index - 1];
+          const feeAdjustedAmount = amountIn
+            .mul(totalFeeRate.denominator.sub(totalFeeRate.numerator))
+            .div(totalFeeRate.denominator);
+          nextBaseValue = price.mul(feeAdjustedAmount).div(U64);
+        } else {
+          const amountOut = index === 0 ? quote.amountOut : acc[index - 1];
+          const feeAdjustedAmount = amountOut.mul(U64).div(price);
+          nextBaseValue = feeAdjustedAmount
+            .mul(totalFeeRate.denominator)
+            .div(totalFeeRate.denominator.sub(totalFeeRate.numerator));
+        }
+
+        acc.push(nextBaseValue);
+        return acc;
+      }, new Array<BN>());
+
+      return acc.add(baseOutputs[baseOutputs.length - 1]);
+    }, new BN(0));
+
+    const totalBaseDecimal = new Decimal(totalBaseValue.toString());
+    const comparableDec = new Decimal(
+      !amountSpecifiedIsInput ? route.totalAmountIn.toString() : route.totalAmountOut.toString()
+    );
+    const priceImpact = amountSpecifiedIsInput
+      ? totalBaseDecimal.sub(comparableDec).div(comparableDec)
+      : comparableDec.sub(totalBaseDecimal).div(comparableDec);
+
```

Comment:
```
const priceImpact = amountSpecifiedIsInput
    ? totalBaseDecimal.sub(comparableDec).div(totalBaseDecimal)
    : comparableDec.sub(totalBaseDecimal).div(comparableDec)
```

## ExactIn (amountSpecifiedIsInput = true)
```math
\begin{array}{rcl}
Out_{best} &=& totalBaseDecimal\\
Out_{estimated} &=& comparableDec\\
Price_{best} &=& \frac{Out_{best}}{In}\\
Price_{estimated} &=& \frac{Out_{estimated}}{In}\\
PriceImpact &=& \frac{Price_{best} - Price_{estimated}}{Price_{best}}\\
&=& \frac{\frac{Out_{best}}{In} - \frac{Out_{estimated}}{In}}{\frac{Out_{best}}{In}}\\
&=& \frac{Out_{best} - Out_{estimated}}{Out_{best}}\\
&=& \frac{totalBaseDecimal - comparableDec}{totalBaseDecimal}\\
\end{array}
```

## ExactOut (amountSpecifiedIsInput = false)
```math
\begin{array}{rcl}
In_{best} &=& totalBaseDecimal\\
In_{estimated} &=& comparableDec\\
Price_{best} &=& \frac{Out}{In_{best}}\\
Price_{estimated} &=& \frac{Out}{In_{estimated}}\\
PriceImpact &=& \frac{Price_{best} - Price_{estimated}}{Price_{best}}\\
&=& \frac{\frac{Out}{In_{best}} - \frac{Out}{In_{estimated}}}{\frac{Out}{In_{best}}}\\
&=& 1 - \frac{In_{best}}{In_{estimated}}\\
&=& \frac{In_{estimated} - In_{best}}{In_{estimated}}\\
&=& \frac{comparableDec - totalBaseDecimal}{comparableDec}
\end{array}
```

Reply:
thank you for clarifying the definition of price impact and verifying the math!